### ISSUE TITLE
Invoke operator lod_tensor_to_array error

### ISSUE THREAD

**Issue by jinxing94** [PaddlePaddle/Paddle#19429]
模型不用optimizer能输出loss,  加上optimizer报错，错误信息如下，求问原因。

  exe.run(program.desc, scope, 0, True, True, fetch_var_name)
paddle.fluid.core_avx.EnforceNotMet: Invoke operator lod_tensor_to_array error.

C++ Callstacks:
Enforce failed. Expected rank_level < x.lod().size(), but received rank_level:0 >= x.lod().size():0.
Input should be a LOD tensor, and size is at least 1 at [/paddle/paddle/fluid/operators/lod_tensor_to_array_op.cc:111]
PaddlePaddle Call Stacks:
0       0x7f69bbf72c88p void paddle::platform::EnforceNotMet::Init<std::string>(std::string, char const*, int) + 360
1       0x7f69bbf72fd7p paddle::platform::EnforceNotMet::EnforceNotMet(std::string const&, char const*, int) + 87
2       0x7f69bc2b04e4p paddle::operators::LoDTensorToArrayOp::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const + 4916
3       0x7f69bded67dcp paddle::framework::OperatorBase::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) + 332
4       0x7f69bc0fd55ep paddle::framework::Executor::RunPreparedContext(paddle::framework::ExecutorPrepareContext*, paddle::framework::Scope*, bool, bool, bool) + 382
5       0x7f69bc1005ffp paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool, std::vector<std::string, std::allocator<std::string> > const&, bool) + 143
6       0x7f69bbf63dedp
7       0x7f69bbfa5076p
8       0x7f6a27c24a02p _PyCFunction_FastCallDict + 386
9       0x7f6a27cb5cb8p
10      0x7f6a27cb8607p _PyEval_EvalFrameDefault + 8023
11      0x7f6a27cb5b54p
12      0x7f6a27cb5e07p
13      0x7f6a27cb93d9p _PyEval_EvalFrameDefault + 11561
14      0x7f6a27cb5b54p
15      0x7f6a27cb5e07p
16      0x7f6a27cb93d9p _PyEval_EvalFrameDefault + 11561
17      0x7f6a27cb5b54p
18      0x7f6a27cb5e07p
19      0x7f6a27cb8607p _PyEval_EvalFrameDefault + 8023
20      0x7f6a27cb5b54p
21      0x7f6a27cb612fp PyEval_EvalCodeEx + 47
22      0x7f6a27cb615bp PyEval_EvalCode + 27
23      0x7f6a27ce95b2p PyRun_FileExFlags + 178
24      0x7f6a27ce9726p PyRun_SimpleFileExFlags + 230
25      0x7f6a27d05d02p Py_Main + 3730
26            0x400ae2p main + 370
27        0x38bfc21b45p __libc_start_main + 245
28            0x400b9dp

**Comment by bjjwwang** [comment]
方便贴一下上下文的代码吗

**Comment by lsabrinax** [comment]
我有一样的问题下面是我的代码
images = fluid.layers.data(name='pixel', shape=data_shape, dtype='float32')
    label_in = fluid.layers.data(
        name='label_in', shape=[1], dtype='int32', lod_level=1)
    label_out = fluid.layers.data(
        name='label_out', shape=[1], dtype='int32', lod_level=1)
    gru_backward, encoded_vector, encoded_proj = encoder_net(images)

    backward_first = fluid.layers.sequence_pool(
        input=gru_backward, pool_type='first')
    decoder_boot = fluid.layers.fc(input=backward_first,
                                   size=decoder_size,
                                   param_attr=clip_attr,
                                   bias_attr=False,
                                   act="relu")
    label_in = fluid.layers.cast(x=label_in, dtype='int64')
    trg_embedding = fluid.layers.embedding(
        input=label_in,
        param_attr=clip_attr,
        size=[num_classes + 2, word_vector_dim],
        dtype='float32')
    prediction = gru_decoder_with_attention(trg_embedding, encoded_vector,
                                            encoded_proj, decoder_boot,
                                            decoder_size, num_classes)
    # fluid.clip.set_gradient_clip(fluid.clip.GradientClipByValue(gradient_clip))
    label_out = fluid.layers.cast(x=label_out, dtype='int64')

    _, maxid = fluid.layers.topk(input=prediction, k=1)
    error_evaluator = fluid.evaluator.EditDistance(
        input=maxid, label=label_out, ignored_tokens=[sos, eos])
    inference_program = fluid.default_main_program().clone(for_test=True)
    fluid.layers.Print(prediction, message='predict', summarize=1)
    fluid.layers.Print(label_out, message='label', summarize=1)
    cost = fluid.layers.softmax_with_cross_entropy(prediction, label=label_out)
    fluid.layers.Print(cost, message='cost', summarize=1)
    sum_cost = fluid.layers.reduce_mean(cost)
LR = args.learning_rate
    if learning_rate_decay == "piecewise_decay":
        learning_rate = fluid.layers.piecewise_decay([150000, 300000, 450000], [LR, LR * 0.1, LR * 0.01, LR * 0.001])
    else:
        learning_rate = LR

    optimizer = fluid.optimizer.Adadelta(
        learning_rate=learning_rate, epsilon=1.0e-8, rho=0.95)
    fluid.layers.Print(learning_rate, message='learning_rate')
    optimizer.minimize(sum_cost)
    fluid.layers.Print(learning_rate, message='learning_rate')
定位到问题就是optimize梯度这里出了问题
下面是gru_decoder_with_attention这个函数的定义
def gru_decoder_with_attention(target_embedding, encoder_vec, encoder_proj,
                               decoder_boot, decoder_size, num_classes):
    def simple_attention(encoder_vec, encoder_proj, decoder_state):
        decoder_state_proj = fluid.layers.fc(input=decoder_state,
                                             size=decoder_size,
                                             param_attr=clip_attr,
                                             bias_attr=False)
        decoder_state_expand = fluid.layers.sequence_expand(
            x=decoder_state_proj, y=encoder_proj)
        concated = encoder_proj + decoder_state_expand
        concated = fluid.layers.tanh(x=concated)
        attention_weights = fluid.layers.fc(input=concated,
                                            size=1,
                                            act=None,
                                            param_attr=clip_attr,
                                            bias_attr=False)
        attention_weights = fluid.layers.sequence_softmax(
            input=attention_weights)
        weigths_reshape = fluid.layers.reshape(x=attention_weights, shape=[-1])
        scaled = fluid.layers.elementwise_mul(
            x=encoder_vec, y=weigths_reshape, axis=0)
        context = fluid.layers.sequence_pool(input=scaled, pool_type='sum')
        return context

    rnn = fluid.layers.DynamicRNN()

    with rnn.block():
        current_word = rnn.step_input(target_embedding)
        encoder_vec = rnn.static_input(encoder_vec)
        encoder_proj = rnn.static_input(encoder_proj)
        hidden_mem = rnn.memory(init=decoder_boot, need_reorder=True)
        context = simple_attention(encoder_vec, encoder_proj, hidden_mem)
        fc_1 = fluid.layers.fc(input=context,
                               size=decoder_size * 3,
                               param_attr=clip_attr,
                               bias_attr=False)
        fc_2 = fluid.layers.fc(input=current_word,
                               size=decoder_size * 3,
                               param_attr=clip_attr,
                               bias_attr=False)
        decoder_inputs = fc_1 + fc_2
        h, _, _ = fluid.layers.gru_unit(
            input=decoder_inputs, hidden=hidden_mem,
            param_attr=clip_attr, bias_attr=clip_attr,
            size=decoder_size * 3)
        rnn.update_memory(hidden_mem, h)
        out = fluid.layers.fc(input=h,
                              size=num_classes + 2,
                              param_attr=clip_attr,
                              bias_attr=clip_attr,
                              act=None)
        rnn.output(out)
    return rnn()
这里是在通过一个RNN网络之后输入softmax_with_cross_entropy。如果我在RNN最后的全连接层直接softmax之后再调用cross_entropy是没有问题的。换成softmax_with_cross_entropy就会报上面的错误信息
